core/src/main/java/jenkins/model/IComputer.java

```diff
@@ -0,0 +1,180 @@
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package jenkins.model;
+
+import edu.umd.cs.findbugs.annotations.NonNull;
+import edu.umd.cs.findbugs.annotations.Nullable;
+import hudson.model.Computer;
+import hudson.model.Node;
+import hudson.security.ACL;
+import hudson.security.AccessControlled;
+import java.util.List;
+import java.util.concurrent.Future;
+import org.jenkins.ui.icon.Icon;
+import org.jenkins.ui.icon.IconSet;
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.Beta;
+
+/**
+ * Interface for computer-like objects meant to be passed to {@code t:executors} tag.
+ *
+ * @since TODO
+ */
+@Restricted(Beta.class)
+public interface IComputer extends AccessControlled {
+    /**
+     * Returns {@link Node#getNodeName() the name of the node}.
+     */
+    @NonNull
+    String getName();
+
+    /**
+     * Used to render the list of executors.
+     * @return a snapshot of the executor display information
+     */
+    @NonNull
+    List<? extends IDisplayExecutor> getDisplayExecutors();
+
+    /**
+     * @return {@code true} if the node is offline. {@code false} if it is online.
+     */
+    boolean isOffline();
+
+    /**
+     * @return the node name for UI purposes.
+     */
+    @NonNull
+    String getDisplayName();
+
+    /**
+     * Returns {@code true} if the computer is accepting tasks. Needed to allow agents programmatic suspension of task
+     * scheduling that does not overlap with being offline.
+     *
+     * @return {@code true} if the computer is accepting tasks
+     * @see hudson.slaves.RetentionStrategy#isAcceptingTasks(Computer)
+     * @see hudson.model.Node#isAcceptingTasks()
+     */
+    boolean isAcceptingTasks();
+
+    /**
+     * @return the URL where to reach specifically this computer, relative to Jenkins URL.
+     */
+    @NonNull
+    String getUrl();
+
+    /**
+     * @return {@code true} if this computer has a defined offline cause, @{code false} otherwise.
+     */
+    default boolean hasOfflineCause() {
+        return getOfflineCauseReason() != null;
+    }
+
+    /**
+     * If the computer was offline (either temporarily or not),
+     * this method will return the cause as a string (without user info).
+     * <p>
+     * {@code hasOfflineCause() == true} implies this must be non-null.
+     *
+     * @return
+     *      empty string if the system was put offline without given a cause.
+     */
+    @Nullable
```

Comment:
I am not sure this annotation even works as documented. I recall a recent case where it did not. I would suggest just using `@CheckForNull`.

Reply:
This is a mistake from my end, actually this is `@NonNull`.